jubilant/_juju.py

```diff
@@ -100,45 +110,83 @@ def deploy(
+        delay: float = 1.0,
+        timeout: float | None = None,
         successes: int = 3,
     ) -> Status:
-        """TODO."""
-        start = time.time()
+        """Wait until ``ready(status)`` returns true.
+
+        This fetches the Juju status repeatedly (waiting *delay* seconds between each call),
+        and returns the last status after the *ready* callable returns true for *successes*
+        times in a row.
+
+        TODO: examples
+
+        Args:
+            ready: Callable that takes a :class:`Status` object and returns true when the wait
+                should be considered ready. It needs to return true *successes* times in a row
+                before ``wait`` returns.
+            error: Callable that takes a :class:`Status` object and returns true when ``wait``
+                should raise an error (:class:`WaitError`).
+            delay: Delay in seconds between status calls.
+            timeout: Overall timeout; :class:`TimeoutError` is raised when this is reached.
+                If not specified, uses the *wait_timeout* specified when the instance was created.
+            successes: Number of times *ready* must return true for the wait to succeed.
+
+        Raises:
+            TimeoutError: If the *timeout* is reached.
+            WaitError: If the *error* callable returns true.
+        """
+        if timeout is None:
+            timeout = self.wait_timeout
+
+        status = None
         success_count = 0
-        this_status = None
-        while time.time() - start < timeout:
-            this_status = self.status(model=model)
-            #        logger.info('wait_status: %s', this_status)  # TODO: ensure better debugging
-            print('TODO wait_status', this_status)
-            if ready_func(this_status):
+        start = time.monotonic()
+
+        while time.monotonic() - start < timeout:
+            prev_status = status
+            status = self.status()
+            if status != prev_status:
+                logger.info('status changed:\n%s', status)
+
+            if error is not None and error(status):
+                msg = f'error function {error.__qualname__} returned false'
+                raise _exception_with_status(WaitError, msg, status)
+
+            if ready(status):
                 success_count += 1
                 if success_count >= successes:
-                    return this_status
+                    return status
             else:
                 success_count = 0
+
             time.sleep(delay)
-        raise Exception(f'timed out after {timeout}, last status:\n{this_status}')
+
+        raise _exception_with_status(TimeoutError, f'timed out after {timeout}s', status)
+
+
+def _exception_with_status(exc_type: type[Exception], msg: str, status: Status | None):
+    if status is None:
+        return exc_type(msg)
+    if hasattr(exc_type, 'add_note'):
+        exc = exc_type(msg)
+        exc.add_note(str(status))
```

Comment:
TIL

Reply:
Yeah, me too! Tony pointed me to the PEP: https://peps.python.org/pep-0678/